### PR TITLE
Make transaction ID required in UnknownTransactionStatusException

### PR DESCRIPTION
**Pull request by brfrn169**
This PR makes the transaction ID required in `UnknownTransactionStatusException`. Please take a look!

### CODE REVIEW

**Review by komamitsu — APPROVED** [review]
LGTM, thanks!

File: core/src/main/java/com/scalar/db/exception/transaction/UnknownTransactionStatusException.java [hunk 1]
```diff
@@ -1,25 +1,20 @@
...
   }
 
   public UnknownTransactionStatusException(String message, Throwable cause, String txId) {
     super(message, cause);
-    this.unknownTxId = Optional.ofNullable(txId);
+    this.unknownTxId = txId;
   }
 
-  public Optional<String> getUnknownTransactionId() {
+  public String getUnknownTransactionId() {
```
Comment (feeblefakie) [inline core/src/main/java/com/scalar/db/exception/transaction/UnknownTransactionStatusException.java:17]: I'm wondering what the effect would be to users.
If an application catches `UnknownTransactionStatusException` and uses `getUnknownTransactionId`  method, then the application has to be modified?

Is it strongly required for future changes?
I'm wondering if the benefit outweighs the downside.

**Review by feeblefakie — COMMENTED** [review]
Thank you!
The code looks good but I left a question for clarification.

File: core/src/main/java/com/scalar/db/exception/transaction/UnknownTransactionStatusException.java (hunk 1 above)
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/exception/transaction/UnknownTransactionStatusException.java:17]: > If an application catches UnknownTransactionStatusException and uses getUnknownTransactionId method, then the application has to be modified?

Yes, it's an incompatible change. So that situation will cause a compile error. But I didn't think that many users actually use `getUnknownTransactionId()` at the moment.

> Is it strongly required for future changes?
> I'm wondering if the benefit outweighs the downside.

I thought I needed this change when I was working on another private project that depends on ScalarDB, and I was wondering why the transaction ID is optional, and when it's null and when it's not. Then, I checked the source code again and found out that we can always return the transaction ID in case of `UnknownTransactionStatusException`. That's the background of this change.

To summarize, the benefit is that the current behavior of the optional transaction ID is a bit confusing for users, and this change fixes this. And the downside is that it's an incompatible change, so if users are using `getUnknownTransactionId`, they have to modify their code when upgrading ScalarDB. And we shouldn't make such an incompatible change between minor versions. But as I said, I didn't think many users were using it at the moment, so I didn't think the impact would be significant.

What do you think?

Comment (komamitsu) [reply core/src/main/java/com/scalar/db/exception/transaction/UnknownTransactionStatusException.java:17]: This might be an off-topic comment, but it would be great if any exception thrown from ScalarDB layer contains a transaction ID information not only in case of `UnknownTransactionStatusException`.

I'll take care of https://github.com/scalar-labs/scalardb-sql/pull/133#discussion_r1135043578, but I'm thinking we might need a consistent way (not only for Spring Data integration 2PC API) to get a transaction ID information from a thrown exception.

Comment (brfrn169) [reply core/src/main/java/com/scalar/db/exception/transaction/UnknownTransactionStatusException.java:17]: As discussed it offline, and we concluded that we don't want to break the backward compatibility so we will leave the original method as it is and add another method that returns non-optional transaction ID.

I will fix this PR. Thanks.

Comment (brfrn169) [reply core/src/main/java/com/scalar/db/exception/transaction/UnknownTransactionStatusException.java]: Fixed in 51311d0bb8642461a63506d2e6d41b166cede76b.

@feeblefakie Can you please take a look again when you have time?

File: core/src/main/java/com/scalar/db/exception/transaction/UnknownTransactionStatusException.java [hunk 2]
```diff
@@ -4,22 +4,33 @@
...
-    this.unknownTxId = Optional.ofNullable(txId);
+    this.unknownTxId = txId;
   }
 
+  /**
+   * @return a transaction ID that associated with the transaction whose status is unknown
+   * @deprecated As of release 3.9.0. Will be removed in release 5.0.0
+   */
+  @Deprecated
   public Optional<String> getUnknownTransactionId() {
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/exception/transaction/UnknownTransactionStatusException.java:24]: To maintain the backward compatibility, we leave this method as it is.

**Review by Torch3333 — APPROVED** [review]
LGTM, thank you!

**Comment by brfrn169** [comment]
@feeblefakie @Torch3333 

I discussed [the comment](https://github.com/scalar-labs/scalardb/pull/829#discussion_r1136437860) with @komamitsu and we came to the conclusion that we should add a transaction ID not only for `UnknownTransactionStatusException` but for all exceptions. So I will close this PR and create another one for that. Thanks for the review.
